### PR TITLE
[Aptos Framework] Add vesting module

### PR DESCRIPTION
**Pull request by movekevin**
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3674)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/vesting.move [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+        // Not used currently but might in the future if this module is upgraded. We want to keep the signer cap just
+        // in case.
+        signer_cap: SignerCapability,
+    }
+
+    /// Resource holding a nonce that's stored at the admin account. We'll use this to create resource accounts for new
+    /// vesting pools so there's no address collision.
+    struct AdminNonce has key {
+        vesting_pool_creation_nonce: u64,
+    }
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/vesting.move:116]: I don't think that this is necessary, we can just use guid if you really need it. Again someone can pre-compute these and create hell for you.

File: aptos-move/framework/aptos-framework/sources/vesting.move [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+            &vesting_schedule,
+            withdrawal_address,
+            vesting_start_timestamp_secs,
+        );
+
+        // If this is the first time this admin account has created a vesting pool, initialize the nonce.
+        let admin_address = signer::address_of(admin);
+        if (!exists<AdminNonce>(admin_address)) {
+            move_to(admin, AdminNonce { vesting_pool_creation_nonce: 0 });
+        };
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/vesting.move:184]: guid please guid

File: aptos-move/framework/aptos-framework/sources/vesting.move [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+        let i = 0;
+        while (i < len) {
+            request_distribution(admin, *vector::borrow(&pool_addresses, i));
+            i = i + 1;
+        };
+    }
+
+    public entry fun request_distribution(admin: &signer, pool_address: address) acquires VestingPool {
+        let vesting_pool = borrow_global_mut<VestingPool>(pool_address);
+        verify_admin(admin, vesting_pool);
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/vesting.move:229]: why do we need admin?

File: aptos-move/framework/aptos-framework/sources/vesting.move [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+        // Request unlock the vested amount + rewards from the stake pool.
+        let total_amount_to_distribute = vested_amount + accumulated_rewards;
+        assert!(total_amount_to_distribute > 0, error::invalid_state(ENO_COINS_TO_DISTRIBUTE));
+        stake::unlock_with_cap(total_amount_to_distribute, &vesting_pool.stake_pool_owner_cap);
+    }
+
+    /// Distribute any withdrawable stake from the stake pool.
+    public entry fun distribute(admin: &signer, pool_address: address) acquires VestingPool {
+        let vesting_pool = borrow_global<VestingPool>(pool_address);
+        verify_admin(admin, vesting_pool);
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/vesting.move:282]: why do we need admin?

File: aptos-move/framework/aptos-framework/sources/vesting.move [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+        assert_active_vesting_pool(vesting_pool);
+
+        // Cannot terminate the vesting pool if there's an in-flight distribution.
+        let (_, withdrawable, _, _pending_unlocked) = stake::get_stake(pool_address);
+        assert!(withdrawable + _pending_unlocked == 0, error::invalid_state(EDISTRIBUTIONS_IN_FLIGHT));
+
+        vesting_pool.state = VESTING_POOL_TERMINATED;
+
+        // Request unlock of all stake.
+        let (active, _, pending_active, _) = stake::get_stake(pool_address);
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/vesting.move:337]: forgive me, but can we not move `pending_active` back?

**Review by davidiw — COMMENTED** [review]
code looks overall fine, but ... where are your tests :O

File: aptos-move/framework/aptos-framework/sources/vesting.move (hunk 1 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/vesting.move]: This is just for convenience as:
1. In genesis or in any scenario where multiple vesting contracts are created in a single tx, account sequence number won't be incremented after each vesting contract creation.
2. Address squatting is a similar challenge to in staking_contract, so create_vesting_contract accepts an optional salt that the caller can pass in to get around address squatting.

File: aptos-move/framework/aptos-framework/sources/vesting.move (hunk 2 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/vesting.move]: Which guid? You mean the event's guid?

File: aptos-move/framework/aptos-framework/sources/vesting.move (hunk 3 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/vesting.move]: We don't. unlock_rewards, vest, and distribute now no longer require the admin

File: aptos-move/framework/aptos-framework/sources/vesting.move (hunk 4 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/vesting.move]: Same

File: aptos-move/framework/aptos-framework/sources/vesting.move (hunk 5 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/vesting.move]: Unfortunately pending_active cannot currently be moved directly to inactive.

File: aptos-move/framework/aptos-framework/sources/vesting.move [hunk 6]
```diff
@@ -0,0 +1,1103 @@
...
+        let grant_pool = &vesting_contract.grant_pool;
+        let shareholders = &pool_u64::shareholders(grant_pool);
+        let len = vector::length(shareholders);
+        let i = 0;
+        while (i < len) {
+            let shareholder = *vector::borrow(shareholders, i);
+            let shares = pool_u64::shares(grant_pool, shareholder);
+            let amount = pool_u64::shares_to_amount_with_total_coins(grant_pool, shares, total_distribution_amount);
+            let share_of_coins = coin::extract(&mut coins, amount);
+            let recipient_address = get_beneficiary(vesting_contract, shareholder);
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/vesting.move:472]: hmm... this whole contract fails if this account is frozen or doesn't exist...

**Review by davidiw — COMMENTED** [review]
overall seems fine... I do think we should have made this more about vesting from something a bit more ... ambiguous... maybe instead of making it pull, we can push into this as the destination for a commissioned contract? that way they can act composably?

File: aptos-move/framework/aptos-framework/sources/vesting.move (hunk 6 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/vesting.move:478]: set_beneficiary checks that the beneficiary account is registered for APT. I can add a nicer error if the account doesn't exist. As for frozen, that's a tricky problem. Admin can always change the beneficiary account though.

File: aptos-move/framework/aptos-framework/sources/vesting.move [hunk 7]
```diff
@@ -0,0 +1,1239 @@
...
+            move_to(admin, AdminStore {
+                vesting_contracts: vector::empty<address>(),
+                nonce: 0,
+                create_events: new_event_handle<CreateVestingContractEvent>(admin),
+            });
+        };
+
+        // Initialize the vesting contract in a new resource account. This allows the same admin to create multiple
+        // pools.
+        let (contract_signer, contract_signer_cap) = create_vesting_contract_account(admin, contract_creation_seed);
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/vesting.move:327]: I still try understand the overview of this contract. what is the main purpose of having this resource account?

File: aptos-move/framework/aptos-framework/sources/vesting.move [hunk 8]
```diff
@@ -0,0 +1,1239 @@
...
+            last_vested_period: 0,
+        }
+    }
+
+    /// Create a vesting contract with a given configurations.
+    public fun create_vesting_contract(
+        admin: &signer,
+        grant: Coin<AptosCoin>,
+        shareholders: &vector<address>,
+        shares: &vector<u64>,
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/vesting.move:278]: I kinda would prefer vector<(address, Coin<AptosCoin>) and we generate the shares here... but I can do that in my code....

**Comment by davidiw** [comment]
I'm ready to approve this PR. It is fairly straightforward. It depends on the commission contract, but that's an independent issue...

**Comment by movekevin** [comment]
> I'm ready to approve this PR. It is fairly straightforward. It depends on the commission contract, but that's an independent issue...

Why is depending on staking_contract module an issue? If not, this vesting module would need to re-implement all the logic regarding stake pool management

File: aptos-move/framework/aptos-framework/sources/vesting.move (hunk 7 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/vesting.move:327]: It isolates the power to control the staking. If we don't create this account and use admin's account directly, the admin would be able to bypass this vesting module and manage the stake themselves, including arbitrarily withdrawing.

File: aptos-move/framework/aptos-framework/sources/vesting.move (hunk 8 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/vesting.move:278]: So you want to pass a vector<Coin> here as shares (and remove the grant argument)? That works too. How strongly do you feel about it? I can make the change if needed as it's fine either way to me.

**Pull request merged**